packages/roosterjs-editor-plugins/lib/CustomReplace/CustomReplace.ts

```diff
@@ -0,0 +1,157 @@
+import { Editor, EditorPlugin } from 'roosterjs-editor-core';
+import { PluginEvent, PluginEventType, PluginKeyPressEvent } from 'roosterjs-editor-types';
+
+type Replacement = {
+    sourceString: string,
+    replacementHTML: string,
+    matchSourceCaseSensitive: boolean
+}
+
+const defaultReplacements: Replacement[] = [
+    {
+        sourceString: ':)',
+        replacementHTML: '🙂',
+        matchSourceCaseSensitive: true,
+    },
+    {
+        sourceString: ';)',
+        replacementHTML: '😉',
+        matchSourceCaseSensitive: true,
+    },
+    {
+        sourceString: ':O',
+        replacementHTML: '😲',
+        matchSourceCaseSensitive: true,
+    },
+    {
+        sourceString: ':o',
+        replacementHTML: '😯',
+        matchSourceCaseSensitive: true,
+    },
+    {
+        sourceString: '<3',
+        replacementHTML: '❤️',
+        matchSourceCaseSensitive: true,
+    },
+]
+
+/**
+ * Wrapper for CustomReplaceContentEditFeature that provides an API for updating the
+ * content edit feature
+ */
+export default class CustomReplacePlugin implements EditorPlugin {
+    private longestReplacementLength: number;
+    private editor: Editor;
+
+    /**
+     * Create instance of CustomReplace plugin
+     * @param features An optional feature set to determine which features the plugin should provide
+     */
+    constructor(private replacements: Replacement[] = defaultReplacements) {
+        this.longestReplacementLength = getLongestReplacementSourceLength(this.replacements);
+    }
+
+    updateReplacements(newReplacements: Replacement[]) {
+        this.replacements = newReplacements;
+        this.longestReplacementLength = getLongestReplacementSourceLength(this.replacements);
+    }
+
+    /**
+     * Get a friendly name of  this plugin
+     */
+    getName() {
+        return 'CustomReplace';
+    }
+
+    /**
+     * Initialize this plugin
+     * @param editor The editor instance
+     */
+    public initialize(editor: Editor): void {
+        this.editor = editor;
+    }
+
+    /**
+     * Dispose this plugin
+     */
+    public dispose(): void {
+    }
+
+    public onPluginEvent(event: PluginEvent) {
+        if (this.editor.isInIME() || event.eventType != PluginEventType.KeyPress) {
+            return;
+        }
+
+        // This handler fires before the event's native behaviour -- defer until after it completes
+        setTimeout(() => this.handlePluginEvent(event), 0)
```

Comment:
1. try editor.runAsync().
2. Suggest pick all the possible last charactor first, then only run this for those last charactors.
For example, for the default replacements, only ")Oo3" can be the last charactor, so we don't need to do any other check if the typing key is not one of them.

Reply:
As I commented in the code, please use input event (for IE, use textinput event instead) so that you don't need this async operation.